### ISSUE TITLE
Reading large data files directly into a database

### ISSUE THREAD

**Issue by coolbutuseless** [hadley/readr#118]
This issue was created after seeing a python/plot.ly example which starts by reading a 3.9GB CSV into a SQLite database: [data analytics with pandas](https://plot.ly/ipython-notebooks/big-data-analytics-with-pandas-and-sqlite/)

The major pain point in repeating this analysis in R, is the initial transfer of the data into a database.

* If there is enough memory, a combination of readr::read_csv() and dplyr::copy_to() would work fine.
* for larger data files e.g. [nyc taxi data](http://www.andresmh.com/nyctaxitrips/) you would have to
    - use something else (e.g. python, or SQLite's CSV import) to do the data transfer
    - write custom R code to read chunks of the CSV and stuff into a DB manually.

The gold standard for doing this in python is probably [blaze odo](https://github.com/ContinuumIO/odo). Which is a graph-based framework for anything-to-anything data manipulation.

Can this data transfer process by made easier with readr? E.g. something like

`readr::read_csv(filename, target=src_sqlite(dbfile))`

**Comment by coolbutuseless** [comment]
An issue with a custom iterative "read a chunk of N rows from CSV, write to db" approach is that column type detection is repeated for each chunk.

So, for the following table, if you read it in chunks of 20

* Read first 20 rows:   `read_csv(file, skip=0, n_max=20)`
* Create database, with name==string, age=integer
* `dbWriteTable(con, tablename, data)`
* Read second 20 rows: `read_csv(file, skip=20, n_max=20)`
* Try to stuff Zohan's age as an integer in the database
    - `dbWriteTable(con, tablename, data, append=TRUE)`
    - and either fail or silently corrupt data

name |  age
---------| -----
adam | 1
betty | 2
...  | ...
yani | 25
zohan | Twenty-Six

**Comment by coolbutuseless** [comment]
I fleshed out a quick example of chunked CSV->sqlite in rstats using readr to check feasability and do some timings: [larger-than-memory CSV to sqlite with readr](https://gist.github.com/coolbutuseless/6697f91ea5bdcc10ad27)

I compared this to a current common method of csv->sqlite import:
```r
dbWriteTable(con, name=table_name, value=csv_file, row.names=FALSE, header=TRUE, sep = ",")
```

The advantages of chunked conversion using `readr`:

* `readr::read_csv` has a faster parser, and overall converted data to SQLite faster.
* `readr` has nicer progress bars and `problem()` output, while a direct dbWriteTable is a bit of a black box.
* `readr::read_csv` correctly handles commas which appear in quoted fields!  \o/
* Easy to swap out `read_csv` for another method like `read_fwf`.
